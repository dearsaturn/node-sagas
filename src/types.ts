import effectBuilder from './effect_builder';
import ApolloClient from 'apollo-client';

export type PutEffect<Payload> = (
    pattern: string,
    payload?: Payload
) => IPutEffectDescription<Payload>;

export type TakeEffect = (pattern: TakePattern) => ITakeEffectDescription;

export type CallEffect<Fn extends (...args: any[]) => any> = (
    callable: (args: Parameters<Fn>) => ReturnType<Fn>,
    args: Parameters<Fn>
) => ICallEffectDescription<Parameters<Fn>, ReturnType<Fn>>;

export interface IActionBuffer<Action extends IAction> {
    isEmpty(): boolean;
    put(action: Action): void;
    take(): Action;
}
export type ActionChannel<Action extends IAction> = (
    pattern: string | Predicate<Action>,
    actionBuffer?: IActionBuffer<Action>
) => IActionChannelEffectDescription<Action, IActionBuffer<Action>>;

export type Predicate<Action extends IAction> = (action: Action) => boolean;

export type TakePattern<
    Action extends IAction = IAction,
    Buffer extends IActionBuffer<Action> = IActionBuffer<Action>
> = string | string[] | Predicate<Action> | IActionChannelEffectDescription<Action, Buffer>;

export type ActionChannelPattern<Action extends IAction> = string | string[] | Predicate<Action>;

export interface IPutEffectDescription<Payload extends {}> extends IEffectDescription {
    pattern: string;
    payload?: Payload;
    kind: 'PUT';
}

export interface ITakeEffectDescription extends IEffectDescription {
    patterns: TakePattern;
    kind: 'TAKE';
}

export interface ICallEffectDescription<Arguments extends any[], CallResponse>
    extends IEffectDescription {
    effect: (...args: Arguments) => CallResponse;
    args: Arguments;
}

export interface IActionChannelEffectDescription<
    Action extends IAction = IAction,
    Buffer extends IActionBuffer<Action> = IActionBuffer<Action>
> extends IEffectDescription {
    pattern: ActionChannelPattern<Action>;
    kind: 'ACTION_CHANNEL';
    buffer: Buffer;
}

export interface IEffectDescription {
    transactionId: string;
    kind: 'PUT' | 'CALL' | 'TAKE' | 'ACTION_CHANNEL';
}

export interface IBaseSagaContext {
    transactionId: string;
    effects: ReturnType<typeof effectBuilder>;
    gqlClient: ApolloClient<any>;
}

export type UnPromisify<T> = T extends Promise<infer U> ? U : T;

type InferActionPayload<T> = T extends Record<keyof T, infer P> ? Record<keyof T, P> : undefined;

export interface IAction<Payload = {}, InferedPayload = InferActionPayload<Payload>> {
    topic: string;
    transactionId: string;
    payload: InferedPayload;
}
