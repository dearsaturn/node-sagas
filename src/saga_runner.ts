import Bluebird from 'bluebird';
import {enums} from '@social-native/snpkg-snapi-authorization';

import {
    IAction,
    IBaseSagaContext,
    IEffectDescription,
    ArrayCombinator,
    RecordCombinator,
    Saga,
    Middleware,
    Next
} from './types';
import {ConsumerMessageBus} from './consumer_message_bus';
import {ProducerMessageBus} from './producer_message_bus';
import {
    isActionChannelEffectDescription,
    isTakeEffectDescription,
    isPutEffectDescription,
    isCallEffectDescription,
    isTakeActionChannelEffectDescription,
    isEffectCombinatorDescription,
    isDelayEffectDescription,
    isGenerator
} from './type_guard';

const {
    WORKER_USER_IDENTITY_HEADER: {WORKER_USER_ID, WORKER_USER_ROLES}
} = enums;

export class SagaRunner<InitialActionPayload, Context extends IBaseSagaContext> {
    protected runEffectWithMiddleware: <EffectDescription extends IEffectDescription>(
        effect: EffectDescription,
        context: Context
    ) => Promise<any>;

    constructor(
        private consumerMessageBus: ConsumerMessageBus,
        private producerMessageBus: ProducerMessageBus,
        middlewares: Array<Middleware<IEffectDescription, Context>> = []
    ) {
        const initialNext: Next<IEffectDescription, Context> = async (effect, ctx) => {
            return this.runEffect(effect, ctx);
        };

        this.runEffectWithMiddleware = middlewares
            .reduceRight((previousNext, middleware: Middleware<IEffectDescription, Context>) => {
                return middleware(previousNext);
            }, initialNext)
            .bind(this);
    }

    public runSaga = async (
        initialAction: IAction<InitialActionPayload>,
        context: Context,
        saga: Saga<InitialActionPayload, Context>
    ): Promise<any> => {
        this.consumerMessageBus.startTransaction(initialAction.transaction_id);
        const result = await this.runGeneratorFsm(saga(initialAction, context), context);
        this.consumerMessageBus.stopTransaction(initialAction.transaction_id);
        return result;
    };

    // tslint:disable-next-line: cyclomatic-complexity
    public runEffect = async <EffectDescription extends IEffectDescription>(
        effectDescription: EffectDescription,
        context: Context
    ) => {
        if (isEffectCombinatorDescription(effectDescription)) {
            const {effects, combinator} = effectDescription;

            if (Array.isArray(effects)) {
                const withRunningEffects: Array<Promise<any>> = effects.map(effect =>
                    this.runEffect(effect, context)
                );

                return await (combinator as ArrayCombinator<IAction>)(withRunningEffects);
            } else if (typeof effects === 'object') {
                const withRunningEffects: Record<string, Promise<any>> = Object.keys(
                    effects
                ).reduce((obj, key) => {
                    return {
                        ...obj,
                        [key]: this.runEffect(effects[key], context)
                    };
                }, {} as Record<string, Promise<any>>);

                return await (combinator as RecordCombinator<IAction>)(withRunningEffects);
            }

            throw new Error(
                'Incompatible effects passed into combinator. Must be an array or object of effects'
            );
        }

        if (isActionChannelEffectDescription(effectDescription)) {
            for (const topic of effectDescription.topics) {
                this.consumerMessageBus.registerTopicObserver({
                    transactionId: effectDescription.transactionId,
                    topic,
                    observer: effectDescription.observer
                });

                await this.consumerMessageBus.streamActionsFromTopic(topic);
            }

            return effectDescription;
        }

        // If this effect already has a stream buffer for events matching the pattern,
        // then just take from the buffer
        if (isTakeActionChannelEffectDescription(effectDescription)) {
            return await effectDescription.buffer.take();
        }

        if (isDelayEffectDescription(effectDescription)) {
            const {delayInMilliseconds, payload} = effectDescription;
            await Bluebird.delay(delayInMilliseconds);
            return payload;
        }

        if (isTakeEffectDescription(effectDescription)) {
            await Bluebird.map(effectDescription.topics, async topic => {
                this.consumerMessageBus.registerTopicObserver({
                    transactionId: effectDescription.transactionId,
                    topic,
                    observer: effectDescription.observer
                });

                await this.consumerMessageBus.streamActionsFromTopic(topic);
            });

            return await effectDescription.buffer.take();
        }

        if (isPutEffectDescription(effectDescription)) {
            const action: IAction<typeof effectDescription.payload> = {
                topic: effectDescription.pattern,
                transaction_id: effectDescription.transactionId,
                payload: effectDescription.payload
            };

            if (
                context.headers &&
                context.headers[WORKER_USER_ID] &&
                context.headers[WORKER_USER_ROLES]
            ) {
                action.userId = context.headers[WORKER_USER_ID];
                action.userRoles = context.headers[WORKER_USER_ROLES].split(',');
            }

            await this.producerMessageBus.putAction(action);

            return;
        }

        if (isCallEffectDescription(effectDescription)) {
            const result = await effectDescription.effect(...(effectDescription.args || []));

            if (isGenerator(result)) {
                return this.runGeneratorFsm(result, context);
            }

            return result;
        }
    };

    protected async runGeneratorFsm<Returned = any | undefined>(
        machine: Generator,
        context: Context,
        {
            previousGeneratorResponse = null,
            didThrow = false
        }: {
            previousGeneratorResponse: any;
            didThrow: boolean;
        } = {previousGeneratorResponse: null, didThrow: false}
    ): Promise<Returned> {
        /**
         * Dereferencing the receiver removes its context, so we need to bind it back to the machine.
         */
        const receiver = didThrow ? machine.throw.bind(machine) : machine.next.bind(machine);

        const {done, value} = receiver(previousGeneratorResponse) as IteratorResult<
            IEffectDescription,
            Returned
        >;

        if (done) {
            return value as Returned;
        }

        try {
            const result = await this.runEffectWithMiddleware(value as IEffectDescription, context);

            return this.runGeneratorFsm(machine, context, {
                previousGeneratorResponse: result,
                didThrow: false
            });
        } catch (error) {
            return this.runGeneratorFsm(machine, context, {
                previousGeneratorResponse: error,
                didThrow: true
            });
        }
    }
}
