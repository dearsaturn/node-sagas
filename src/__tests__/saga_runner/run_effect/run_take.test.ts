import {SagaRunner} from '../../../saga_runner';
import {withTopicCleanup, seedTopic} from '../../kafka_utils';
import {runnerUtilityFactory} from '../runner_utility_factory';
import {DEFAULT_TEST_TIMEOUT} from '../../constants';
import {IAction, ActionChannelInput, IActionChannelEffectDescription} from '../../../types';

describe(SagaRunner.name, function() {
    describe('take', function() {
        describe('given simple patterns', function() {
            const simplePatterns: Array<{
                patternKind: string;
                topics: string[];
                pattern: ActionChannelInput<IAction<{bart_simpson: string}>>;
            }> = [
                {
                    patternKind: 'string',
                    topics: ['test-topic-1'],
                    pattern: 'test-topic-1'
                },
                {
                    patternKind: 'string[]',
                    topics: ['test-topic-1', 'test-topic-2'],
                    pattern: ['test-topic-1', 'test-topic-2']
                },
                {
                    patternKind: 'PredicateRecord',
                    topics: ['test-topic-1'],
                    pattern: {
                        pattern: 'test-topic-1',
                        predicate: () => true
                    }
                }
            ];

            for (const {patternKind, topics, pattern} of simplePatterns) {
                describe(`given a pattern of kind ${patternKind}`, function() {
                    it(
                        'begins streaming actions from the stream matching the pattern',
                        async function() {
                            await withTopicCleanup(topics)(async () => {
                                const runnerUtils = await runnerUtilityFactory();
                                const {runner, effectBuilder, spy, context} = runnerUtils;
                                const observerRegisteredSpy = spy.consumer('registerTopicObserver');
                                const streamActionsSpy = spy.consumer('streamActionsFromTopic');

                                /** This will execute after the runEffect promise is initiated. */

                                setTimeout(async () => {
                                    await seedTopic(topics[0], [
                                        {
                                            transaction_id: runnerUtils.transactionId,
                                            payload: {
                                                bart_simpson: 'eat_my_shorts'
                                            }
                                        }
                                    ]);
                                }, 5000);

                                const payload = await runner.runEffect(
                                    effectBuilder.take(pattern),
                                    context
                                );

                                expect(payload).toMatchSnapshot();
                                expect(observerRegisteredSpy.mock.calls).toMatchSnapshot();
                                expect(streamActionsSpy.mock.calls).toMatchSnapshot();

                                await runnerUtils.closeBuses();
                            });
                        },
                        DEFAULT_TEST_TIMEOUT * 2
                    );
                });
            }
        });

        it(
            'races results from the streams',
            async function() {
                await withTopicCleanup(['test-take-slow', 'test-take-fast'])(
                    async ([slow, fast]) => {
                        const runnerUtils = await runnerUtilityFactory();
                        const {runner, effectBuilder, spy, context} = runnerUtils;
                        const observerRegisteredSpy = spy.consumer('registerTopicObserver');
                        const streamActionsSpy = spy.consumer('streamActionsFromTopic');

                        /** This will execute after the runEffect promise is initiated. */

                        setTimeout(async () => {
                            await seedTopic(slow, [
                                {
                                    transaction_id: runnerUtils.transactionId,
                                    payload: {
                                        bart_simpson: 'loser'
                                    }
                                }
                            ]);
                        }, 3000);

                        setTimeout(async () => {
                            await seedTopic(fast, [
                                {
                                    transaction_id: runnerUtils.transactionId,
                                    payload: {
                                        bart_simpson: 'winner'
                                    }
                                }
                            ]);
                        }, 1000);

                        const payload = await runner.runEffect(
                            effectBuilder.take([slow, fast]),
                            context
                        );

                        expect(payload).toMatchSnapshot();
                        expect(observerRegisteredSpy.mock.calls).toMatchSnapshot();
                        expect(streamActionsSpy.mock.calls).toMatchSnapshot();

                        await runnerUtils.closeBuses();
                    }
                );
            },
            DEFAULT_TEST_TIMEOUT * 2
        );

        describe('given an action channel pattern', function() {
            it(
                "takes from the action channel's buffer",
                async function() {
                    await withTopicCleanup(['test-take-action-channel'])(async ([topic]) => {
                        const runnerUtils = await runnerUtilityFactory();
                        const {runner, effectBuilder, spy, context} = runnerUtils;
                        const observerRegisteredSpy = spy.consumer('registerTopicObserver');
                        const streamActionsSpy = spy.consumer('streamActionsFromTopic');

                        setTimeout(async () => {
                            await seedTopic(topic, [
                                {
                                    transaction_id: runnerUtils.transactionId,
                                    payload: {
                                        bart_simpson: 'bad'
                                    }
                                }
                            ]);
                        }, 2000);

                        const channel: IActionChannelEffectDescription<IAction<{
                            bart_simpson: string;
                        }>> = await runner.runEffect(effectBuilder.actionChannel(topic), context);

                        const bufferSpy = jest.spyOn(channel.buffer, 'put');

                        const payload = await runner.runEffect(
                            effectBuilder.take(channel),
                            context
                        );

                        expect(payload).toMatchSnapshot();
                        expect(observerRegisteredSpy.mock.calls).toMatchSnapshot();
                        expect(streamActionsSpy.mock.calls).toMatchSnapshot();
                        expect(bufferSpy.mock.calls).toMatchSnapshot();

                        await runnerUtils.closeBuses();
                    });
                },
                DEFAULT_TEST_TIMEOUT
            );
        });
    });
});
