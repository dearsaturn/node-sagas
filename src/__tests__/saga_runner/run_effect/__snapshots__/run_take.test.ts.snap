// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SagaRunner take given an action channel pattern takes from the action channel's buffer 1`] = `
Object {
  "headers": Object {},
  "payload": Object {
    "bart_simpson": "bad",
  },
  "topic": "test-take-action-channel",
  "transaction_id": "static-transaction-id",
}
`;

exports[`SagaRunner take given an action channel pattern takes from the action channel's buffer 2`] = `
Array [
  Array [
    Object {
      "observer": [Function],
      "topic": "test-take-action-channel",
      "transactionId": "static-transaction-id",
    },
  ],
]
`;

exports[`SagaRunner take given an action channel pattern takes from the action channel's buffer 3`] = `
Array [
  Array [
    "test-take-action-channel",
  ],
]
`;

exports[`SagaRunner take given an action channel pattern takes from the action channel's buffer 4`] = `
Array [
  Array [
    Object {
      "headers": Object {},
      "payload": Object {
        "bart_simpson": "bad",
      },
      "topic": "test-take-action-channel",
      "transaction_id": "static-transaction-id",
    },
  ],
]
`;

exports[`SagaRunner take given simple patterns given a pattern of kind PredicateRecord begins streaming actions from the stream matching the pattern 1`] = `
Object {
  "headers": Object {},
  "payload": Object {
    "bart_simpson": "eat_my_shorts",
  },
  "topic": "test-topic-1",
  "transaction_id": "static-transaction-id",
}
`;

exports[`SagaRunner take given simple patterns given a pattern of kind PredicateRecord begins streaming actions from the stream matching the pattern 2`] = `
Array [
  Array [
    Object {
      "observer": [Function],
      "topic": "test-topic-1",
      "transactionId": "static-transaction-id",
    },
  ],
]
`;

exports[`SagaRunner take given simple patterns given a pattern of kind PredicateRecord begins streaming actions from the stream matching the pattern 3`] = `
Array [
  Array [
    "test-topic-1",
  ],
]
`;

exports[`SagaRunner take given simple patterns given a pattern of kind string begins streaming actions from the stream matching the pattern 1`] = `
Object {
  "headers": Object {},
  "payload": Object {
    "bart_simpson": "eat_my_shorts",
  },
  "topic": "test-topic-1",
  "transaction_id": "static-transaction-id",
}
`;

exports[`SagaRunner take given simple patterns given a pattern of kind string begins streaming actions from the stream matching the pattern 2`] = `
Array [
  Array [
    Object {
      "observer": [Function],
      "topic": "test-topic-1",
      "transactionId": "static-transaction-id",
    },
  ],
]
`;

exports[`SagaRunner take given simple patterns given a pattern of kind string begins streaming actions from the stream matching the pattern 3`] = `
Array [
  Array [
    "test-topic-1",
  ],
]
`;

exports[`SagaRunner take given simple patterns given a pattern of kind string[] begins streaming actions from the stream matching the pattern 1`] = `
Object {
  "headers": Object {},
  "payload": Object {
    "bart_simpson": "eat_my_shorts",
  },
  "topic": "test-topic-1",
  "transaction_id": "static-transaction-id",
}
`;

exports[`SagaRunner take given simple patterns given a pattern of kind string[] begins streaming actions from the stream matching the pattern 2`] = `
Array [
  Array [
    Object {
      "observer": [Function],
      "topic": "test-topic-1",
      "transactionId": "static-transaction-id",
    },
  ],
  Array [
    Object {
      "observer": [Function],
      "topic": "test-topic-2",
      "transactionId": "static-transaction-id",
    },
  ],
]
`;

exports[`SagaRunner take given simple patterns given a pattern of kind string[] begins streaming actions from the stream matching the pattern 3`] = `
Array [
  Array [
    "test-topic-1",
  ],
  Array [
    "test-topic-2",
  ],
]
`;

exports[`SagaRunner take races results from the streams 1`] = `
Object {
  "headers": Object {},
  "payload": Object {
    "bart_simpson": "winner",
  },
  "topic": "test-take-fast",
  "transaction_id": "static-transaction-id",
}
`;

exports[`SagaRunner take races results from the streams 2`] = `
Array [
  Array [
    Object {
      "observer": [Function],
      "topic": "test-take-slow",
      "transactionId": "static-transaction-id",
    },
  ],
  Array [
    Object {
      "observer": [Function],
      "topic": "test-take-fast",
      "transactionId": "static-transaction-id",
    },
  ],
]
`;

exports[`SagaRunner take races results from the streams 3`] = `
Array [
  Array [
    "test-take-slow",
  ],
  Array [
    "test-take-fast",
  ],
]
`;
